homework1.py

```diff
@@ -0,0 +1,134 @@
+from typing import List
+from pydantic import BaseModel
+from abc import ABC, abstractmethod
+
+# Pydantic модель для книги
+class BookModel(BaseModel):
+    title: str
+    author: str
+    year: int
+
+    def __str__(self):
+        return f"{self.title} ({self.author}, {self.year})"
+
+#Абстрактний клас
+class AbstractLiterature(ABC):
+    @abstractmethod
+    def litinfo(self):
+        raise NotImplementedError
+
+# Клас "Книга" з методом для повернення інформації про книгу
+class Book(AbstractLiterature):
+    def __init__(self, book_model: BookModel):
+        self.book_model = book_model
+
+    def __str__(self):
+        return str(self.book_model)
+
+    #Імплементація абстрактного методу
+    def litinfo(self):
+        return self.__str__()
+
+# Клас "Журнал", який наслідується від "Книги"
+class Journal(Book):
+    pass
+# Клас "Бібліотека" з атрибутом у вигляді списку книг (зроблений приватним)
+class Library:
+    def __init__(self):
+        self.__books: List[Book] = []
+
+    # Функції для перегляду і взаємодії з приватним класом
+    def get_books(self):
+        return self.__books
+
+    def set_books(self, books):
+        self.__books = books
+
+
+    # Декоратор, який пише лог при додаванні нової книги у бібліотеку
+    def add_log(func):
+        def wrapper(self, book):
+            print(f"\nAdding the book: {book}")
+            result = func(self, book)
+            print("\nList of books in the library after addition:")
+            for b in self.get_books():
+                print(b)
+            return result
+
+        return wrapper
+
+    # Декоратор, який перевіряє наявність книги у бібліотеці перед видаленням
+    def check_existence(func):
+        def wrapper(self, book):
+            if book in self.get_books():
+                print(f"\nRemoving the book '{book}'")
+                result = func(self, book)
+            else:
+                print(f"Book '{book}' has not been found in the library")
+                result = None
+            return result
+
+        return wrapper
+
+    @add_log
+    def add_book(self, book: Book):
+        self.get_books().append(book)
+
+    @check_existence
+    def remove_book(self, book: Book):
+        self.get_books().remove(book)
+
+    # Ітератор, який дозволяє проходитися по всіх книгах у бібліотеці
+    def __iter__(self):
+        return iter(self.get_books())
```

Comment:
This is iterator by list of books, but I meant your own custom iterator class with `__iter__` and `__next__` methods. I'll close your homework, but please make changes for it)